pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -237,28 +243,89 @@ public Object visit(ASTName node, Object data) {
          * for memory - TODO: Investigate if there is a valid memory concern or
          * not
          */
-        if (node.getNameDeclaration() == null) {
-            // Skip these scenarios as there is no type to populate in these
-            // cases:
-            // 1) Parent is a PackageDeclaration, which is not a type
-            // 2) Parent is a ImportDeclaration, this is handled elsewhere.
-            if (!(node.jjtGetParent() instanceof ASTPackageDeclaration
-                    || node.jjtGetParent() instanceof ASTImportDeclaration)) {
-                String name = node.getImage();
-                if (name.indexOf('.') != -1) {
-                    name = name.substring(0, name.indexOf('.'));
-                }
-                populateType(node, name);
-            }
-        } else {
+
+
+        Class accessingClass = getEnclosingTypeDeclaration(node);
+
+        String[] dotSplitImage = node.getImage().split("\\.");
+        Class previousNameType = getVariableNameType(node.getScope(), dotSplitImage[0], accessingClass);
+
+        if (node.getNameDeclaration() != null && previousNameType == null) {
             // Carry over the type from the declaration
             if (node.getNameDeclaration().getNode() instanceof TypeNode) {
                 node.setType(((TypeNode) node.getNameDeclaration().getNode()).getType());
             }
         }
+
+
+        // Skip these scenarios as there is no type to populate in these
```

Comment:
TODO (separate PR), when visiting those nodes, don't call super, so we won't visit these children. By trimming the tree we get a little performance boost and can't remove this check simplifying the code. Document properly why you don't call super on those nodes.